File: android/app/src/main/java/com/bensbible/app/viewmodel/ReaderViewModel.kt
package com.bensbible.app.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bensbible.app.data.AnnotationRepository
import com.bensbible.app.data.BibleDataService
import com.bensbible.app.data.VerseAnnotationEntity
import com.bensbible.app.model.BibleLocation
import com.bensbible.app.model.Chapter
import com.bensbible.app.model.HighlightColor
import com.bensbible.app.model.Verse
import com.bensbible.app.model.VerseID
import kotlinx.coroutines.launch

class ReaderViewModel(
    private val dataService: BibleDataService,
    private val repository: AnnotationRepository
) : ViewModel() {

    var bookNames by mutableStateOf<List<String>>(emptyList())
        private set

    var currentLocation by mutableStateOf(BibleLocation.genesis1)
        private set

    var currentChapter by mutableStateOf<Chapter?>(null)
        private set

    var currentBookChapterCount by mutableIntStateOf(0)
        private set

    var selectedVerseIDs by mutableStateOf<Set<VerseID>>(emptySet())
        private set

    val chapterAnnotations = mutableStateMapOf<String, VerseAnnotationEntity>()

    var isPickerPresented by mutableStateOf(false)
    var isHighlightPickerPresented by mutableStateOf(false)
    var isShareSheetPresented by mutableStateOf(false)
    var isNoteEditorPresented by mutableStateOf(false)
    var noteEditingVerseID by mutableStateOf<VerseID?>(null)
        private set
    var noteEditingText by mutableStateOf("")
    var scrollToVerseID by mutableStateOf<String?>(null)
    var highlightedVerseID by mutableStateOf<String?>(null)

    val hasSelection: Boolean get() = selectedVerseIDs.isNotEmpty()
    val selectedCount: Int get() = selectedVerseIDs.size

    val selectedVerseTexts: List<Pair<Int, String>>
        get() {
            val chapter = currentChapter ?: return emptyList()
            val sorted = selectedVerseIDs.sorted()
            return sorted.mapNotNull { id ->
                chapter.verses.find { it.number == id.verse }?.let { verse ->
                    verse.number to verse.text
                }
            }
        }

    val selectedVerseReference: String
        get() = VerseID.displayRange(selectedVerseIDs)

    val canGoNext: Boolean
        get() {
            if (currentLocation.chapterNumber < currentBookChapterCount) return true
            return bookAfter(currentLocation.bookName) != null
        }

    val canGoPrevious: Boolean
        get() {
            if (currentLocation.chapterNumber > 1) return true
            return bookBefore(currentLocation.bookName) != null
        }

    fun onAppear() {
        if (bookNames.isNotEmpty()) return
        try {
            bookNames = dataService.loadBookNames()
            loadCurrentChapter()
        } catch (e: Exception) {
            // Log error
        }
    }

    fun navigateTo(book: String, chapter: Int, verse: Int? = null) {
        currentLocation = BibleLocation(bookName = book, chapterNumber = chapter, verseNumber = verse)
        deselectAll()
        try {
            loadCurrentChapter()
            if (verse != null) {
                scrollToVerseID = verse.toString()
                highlightedVerseID = verse.toString()
            }
        } catch (e: Exception) {
            // Log error
        }
    }

    fun nextChapter() {
        if (currentLocation.chapterNumber < currentBookChapterCount) {
            navigateTo(currentLocation.bookName, currentLocation.chapterNumber + 1)
        } else {
            bookAfter(currentLocation.bookName)?.let { navigateTo(it, 1) }
        }
    }

    fun previousChapter() {
        if (currentLocation.chapterNumber > 1) {
            navigateTo(currentLocation.bookName, currentLocation.chapterNumber - 1)
        } else {
            bookBefore(currentLocation.bookName)?.let { prev ->
                val count = chapterCount(prev)
                navigateTo(prev, count)
            }
        }
    }

    fun chapterCount(bookName: String): Int {
        return try {
            dataService.loadBook(bookName).chapters.size
        } catch (e: Exception) {
            0
        }
    }

    fun verseID(verse: Verse): VerseID {
        return VerseID(
            book = currentLocation.bookName,
            chapter = currentLocation.chapterNumber,
            verse = verse.number
        )
    }

    fun toggleVerseSelection(verse: Verse) {
        val id = verseID(verse)
        selectedVerseIDs = if (selectedVerseIDs.contains(id)) {
            selectedVerseIDs - id
        } else {
            selectedVerseIDs + id
        }
    }

    fun isSelected(verse: Verse): Boolean {
        return selectedVerseIDs.contains(verseID(verse))
    }

    fun deselectAll() {
        selectedVerseIDs = emptySet()
    }

    fun annotation(verse: Verse): VerseAnnotationEntity? {
        return chapterAnnotations[verseID(verse).key]
    }

    fun applyHighlight(color: HighlightColor?) {
        if (selectedVerseIDs.isEmpty()) return
        val ids = selectedVerseIDs.toList()
        viewModelScope.launch {
            repository.setHighlight(color, ids)
            loadAnnotationsForCurrentChapter()
        }
        deselectAll()
    }

    fun bookmarkSelectedVerses() {
        if (selectedVerseIDs.isEmpty()) return
        val ids = selectedVerseIDs.toList()
        viewModelScope.launch {
            repository.toggleBookmark(ids)
            loadAnnotationsForCurrentChapter()
        }
        deselectAll()
    }

    fun beginNoteEditing() {
        val firstID = selectedVerseIDs.minByOrNull { it.verse } ?: return
        noteEditingVerseID = firstID
        noteEditingText = chapterAnnotations[firstID.key]?.noteText ?: ""
        isNoteEditorPresented = true
    }

    fun saveNote() {
        val verseID = noteEditingVerseID ?: return
        val text = noteEditingText
        viewModelScope.launch {
            repository.saveNote(text, verseID)
            loadAnnotationsForCurrentChapter()
        }
        noteEditingVerseID = null
        noteEditingText = ""
        deselectAll()
    }

    fun cancelNoteEditing() {
        noteEditingVerseID = null
        noteEditingText = ""
    }

    fun loadAnnotationsForCurrentChapter() {
        viewModelScope.launch {
            val annotations = repository.fetchAnnotations(
                currentLocation.bookName,
                currentLocation.chapterNumber
            )
            chapterAnnotations.clear()
            chapterAnnotations.putAll(annotations)
        }
    }

    private fun loadCurrentChapter() {
        val chapter = dataService.loadChapter(
            currentLocation.bookName,
            currentLocation.chapterNumber
        )
        currentChapter = chapter
        val book = dataService.loadBook(currentLocation.bookName)
        currentBookChapterCount = book.chapters.size
        loadAnnotationsForCurrentChapter()
    }

    private fun bookAfter(name: String): String? {
        val index = bookNames.indexOf(name)
        return if (index >= 0 && index + 1 < bookNames.size) bookNames[index + 1] else null
    }

    private fun bookBefore(name: String): String? {
        val index = bookNames.indexOf(name)
        return if (index > 0) bookNames[index - 1] else null
    }
}

File: BensBible2/ViewModels/ReaderViewModel.swift
import Foundation
import SwiftUI

@Observable
final class ReaderViewModel {
    private let dataService: BibleDataService

    var bookNames: [String] = []
    var currentLocation: BibleLocation = .genesis1
    var currentChapter: Chapter?
    var currentBookChapterCount: Int = 0
    var isPickerPresented = false
    var scrollToVerseID: String?
    var highlightedVerseID: String?
    var errorMessage: String?

    // MARK: - Selection & Annotations

    var selectedVerseIDs: Set<VerseID> = []
    var chapterAnnotations: [String: VerseAnnotation] = [:]
    var annotationService: AnnotationService?

    var isHighlightPickerPresented = false
    var isNoteEditorPresented = false
    var isShareSheetPresented = false
    var noteEditingVerseID: VerseID?
    var noteEditingText: String = ""

    var hasSelection: Bool { !selectedVerseIDs.isEmpty }
    var selectedCount: Int { selectedVerseIDs.count }

    var selectedVerseTexts: [(number: Int, text: String)] {
        guard let chapter = currentChapter else { return [] }
        let selectedNumbers = selectedVerseIDs.map(\.verse)
        return chapter.verses
            .filter { selectedNumbers.contains($0.number) }
            .sorted { $0.number < $1.number }
            .map { (number: $0.number, text: $0.text) }
    }

    var selectedVerseReference: String {
        VerseID.displayRange(from: selectedVerseIDs)
    }

    init(dataService: BibleDataService = LocalBibleDataService()) {
        self.dataService = dataService
    }

    func configure(annotationService: AnnotationService) {
        self.annotationService = annotationService
    }

    func onAppear() {
        guard bookNames.isEmpty else { return }
        do {
            bookNames = try dataService.loadBookNames()
            try loadCurrentChapter()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func navigateTo(book: String, chapter: Int, verse: Int? = nil) {
        currentLocation = BibleLocation(bookName: book, chapterNumber: chapter, verseNumber: verse)
        deselectAll()
        do {
            try loadCurrentChapter()
            if let verse {
                scrollToVerseID = String(verse)
                highlightedVerseID = String(verse)
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func nextChapter() {
        if currentLocation.chapterNumber < currentBookChapterCount {
            navigateTo(book: currentLocation.bookName, chapter: currentLocation.chapterNumber + 1)
        } else if let nextBook = bookAfter(currentLocation.bookName) {
            navigateTo(book: nextBook, chapter: 1)
        }
    }

    func previousChapter() {
        if currentLocation.chapterNumber > 1 {
            navigateTo(book: currentLocation.bookName, chapter: currentLocation.chapterNumber - 1)
        } else if let prevBook = bookBefore(currentLocation.bookName) {
            let count = chapterCount(for: prevBook)
            navigateTo(book: prevBook, chapter: count)
        }
    }

    var canGoNext: Bool {
        if currentLocation.chapterNumber < currentBookChapterCount { return true }
        return bookAfter(currentLocation.bookName) != nil
    }

    var canGoPrevious: Bool {
        if currentLocation.chapterNumber > 1 { return true }
        return bookBefore(currentLocation.bookName) != nil
    }

    func chapterCount(for bookName: String) -> Int {
        do {
            let book = try dataService.loadBook(named: bookName)
            return book.chapters.count
        } catch {
            return 0
        }
    }

    // MARK: - Verse Selection

    func verseID(for verse: Verse) -> VerseID {
        VerseID(book: currentLocation.bookName, chapter: currentLocation.chapterNumber, verse: verse.number)
    }

    func toggleVerseSelection(_ verse: Verse) {
        let id = verseID(for: verse)
        if selectedVerseIDs.contains(id) {
            selectedVerseIDs.remove(id)
        } else {
            selectedVerseIDs.insert(id)
        }
    }

    func isSelected(_ verse: Verse) -> Bool {
        selectedVerseIDs.contains(verseID(for: verse))
    }

    func deselectAll() {
        selectedVerseIDs.removeAll()
    }

    func annotation(for verse: Verse) -> VerseAnnotation? {
        let id = verseID(for: verse)
        return chapterAnnotations[id.key]
    }

    // MARK: - Annotation Actions

    func applyHighlight(_ color: HighlightColor?) {
        guard !selectedVerseIDs.isEmpty else { return }
        annotationService?.setHighlight(color: color, for: Array(selectedVerseIDs))
        loadAnnotationsForCurrentChapter()
        deselectAll()
    }

    func bookmarkSelectedVerses() {
        guard !selectedVerseIDs.isEmpty else { return }
        annotationService?.toggleBookmark(for: Array(selectedVerseIDs))
        loadAnnotationsForCurrentChapter()
        deselectAll()
    }

    func beginNoteEditing() {
        guard let firstID = selectedVerseIDs.sorted(by: { $0.verse < $1.verse }).first else { return }
        noteEditingVerseID = firstID
        noteEditingText = chapterAnnotations[firstID.key]?.noteText ?? ""
        isNoteEditorPresented = true
    }

    func saveNote() {
        guard let verseID = noteEditingVerseID else { return }
        annotationService?.saveNote(text: noteEditingText, for: verseID)
        loadAnnotationsForCurrentChapter()
        noteEditingVerseID = nil
        noteEditingText = ""
        deselectAll()
    }

    func cancelNoteEditing() {
        noteEditingVerseID = nil
        noteEditingText = ""
    }

    // MARK: - Private

    private func loadCurrentChapter() throws {
        let chapter = try dataService.loadChapter(
            bookName: currentLocation.bookName,
            chapter: currentLocation.chapterNumber
        )
        currentChapter = chapter
        let book = try dataService.loadBook(named: currentLocation.bookName)
        currentBookChapterCount = book.chapters.count
        loadAnnotationsForCurrentChapter()
    }

    func loadAnnotationsForCurrentChapter() {
        guard let service = annotationService else { return }
        chapterAnnotations = service.fetchAnnotations(
            book: currentLocation.bookName,
            chapter: currentLocation.chapterNumber
        )
    }

    private func bookAfter(_ name: String) -> String? {
        guard let index = bookNames.firstIndex(of: name),
              index + 1 < bookNames.count else { return nil }
        return bookNames[index + 1]
    }

    private func bookBefore(_ name: String) -> String? {
        guard let index = bookNames.firstIndex(of: name),
              index > 0 else { return nil }
        return bookNames[index - 1]
    }
}
